utils/createhttpclient.go

```diff
@@ -2,14 +2,33 @@ package utils
 
 import (
 	"crypto/tls"
+	"crypto/x509"
 	"net/http"
 	"os"
 	"strings"
 )
 
 func CreateHTTPClient() *http.Client {
 	skipVerify := strings.ToLower(os.Getenv("GOTIFY_SKIP_VERIFY_TLS")) == "true"
+	certFile, hasCA := os.LookupEnv("GOTIFY_TLS_ROOT_CA")
+
 	customTransport := http.DefaultTransport.(*http.Transport).Clone()
-	customTransport.TLSClientConfig = &tls.Config{InsecureSkipVerify: skipVerify}
+	rootCAs := customTransport.TLSClientConfig.RootCAs
+	if hasCA {
+		cert, err := os.ReadFile(certFile)
+		if err != nil {
+			Exit1With("Failed to read CA file:", err)
+		}
+		rootCAs = x509.NewCertPool()
+		ok := rootCAs.AppendCertsFromPEM(cert)
+		if !ok {
+			Exit1With("Failed to parse CA file", certFile)
+		}
+	}
```

Comment:
Not 100% necessary but would be good to have a test on security critical logic.

Reply:
I would just write the PEM files as a string.  You can also generate them on the fly and use net.Pipe() to simulate a network connection internally. here's a demo: 

<details><summary>Code</summary>

```go
package main

import (
	"context"
	"crypto/ed25519"
	"crypto/rand"
	"crypto/tls"
	"crypto/x509"
	"encoding/pem"
	"errors"
	"log"
	"math/big"
	"net"
	"net/http"
	"sync"
	"sync/atomic"
	"time"
)

func newCA() ([]byte, func(domain string) (ed25519.PublicKey, ed25519.PrivateKey)) {
	caPubKey, caPrivKey, err := ed25519.GenerateKey(rand.Reader)
	if err != nil {
		log.Panicf("failed to generate key: %v", err)
	} 
	cert, err := x509.CreateCertificate(rand.Reader, &x509.Certificate{
		SerialNumber:          big.NewInt(1),
		NotAfter:              time.Now().Add(time.Hour),
		IsCA:                  true,
		BasicConstraintsValid: true,
	}, &x509.Certificate{}, caPubKey, caPrivKey)

	if err != nil {
		log.Panicf("failed to create certificate: %v", err)
	}

	certParsed, err := x509.ParseCertificate(cert)
	if err != nil {
		log.Panicf("failed to parse certificate: %v", err)
	}

	return pem.EncodeToMemory(&pem.Block{Type: "CERTIFICATE", Bytes: cert}), func(domain string) (ed25519.PublicKey, ed25519.PrivateKey) {
		pubKey, privKey, err := ed25519.GenerateKey(rand.Reader)
		if err != nil {
			log.Panicf("failed to generate key: %v", err)
		}

		cert, err := x509.CreateCertificate(rand.Reader, &x509.Certificate{
			DNSNames:     []string{domain},
			SerialNumber: big.NewInt(2),
			NotAfter:     time.Now().Add(time.Hour),
		}, certParsed, pubKey, caPrivKey)

		if err != nil {
			log.Panicf("failed to create certificate: %v", err)
		}

		privPEM, err := x509.MarshalPKCS8PrivateKey(privKey)
		if err != nil {
			log.Panicf("failed to marshal private key: %v", err)
		}

		return pem.EncodeToMemory(
				&pem.Block{Type: "CERTIFICATE", Bytes: cert}),
			pem.EncodeToMemory(
				&pem.Block{Type: "PRIVATE KEY", Bytes: privPEM})
	}
}

func main() {
	caPEM, signer := newCA()
	wrongCAPEM, wrongSigner := newCA()

	certPEM, certPriv := signer("gotify.local")
	wrongDomainPEM, wrongDomainPriv := signer("gotify.invalid")
	wrongCAPEM, wrongCAPriv := wrongSigner("gotify.local")

	testTrust := func(clientFactory func(conn net.Conn) *http.Client, serverPEM []byte, serverKey []byte) bool {
		serverSide, clientSide := net.Pipe()

		serverCert, err := tls.X509KeyPair(serverPEM, serverKey)
		if err != nil {
			panic(err)
		}

		tlsServer := tls.Server(serverSide, &tls.Config{
			Certificates: []tls.Certificate{
				serverCert,
			},
		})
		client := clientFactory(clientSide)

		var failed uint32

		wg := sync.WaitGroup{}
		wg.Add(2)

		go func() {
			defer serverSide.Close()
			defer wg.Done()

			if err := tlsServer.Handshake(); err == nil {
				tlsServer.Write([]byte("HTTP/1.1 200 OK\r\n\r\n"))
			}
		}()

		go func() {
			defer clientSide.Close()
			defer wg.Done()

			if _, err := client.Get("https://gotify.local"); err != nil {
				if _, ok := errors.Unwrap(err).(*tls.CertificateVerificationError); ok {
					atomic.StoreUint32(&failed, 1)
				} else {
					log.Panicf("client handshake failed: %v", err)
				}
			}
		}()

		wg.Wait()

		return atomic.LoadUint32(&failed) == 0
	}

	certPool := x509.NewCertPool()
	if !certPool.AppendCertsFromPEM(caPEM) {
		panic("failed to append CA certificate")
	}

	clientFactory := func(c net.Conn) *http.Client {
		return &http.Client{
			Transport: &http.Transport{
				DialContext: func(_ context.Context, network, addr string) (net.Conn, error) {
					return c, nil
				},
				TLSClientConfig: &tls.Config{
					RootCAs: certPool,
				},
			},
		}
	}

	mitmClientFactory := func(c net.Conn) *http.Client {
		return &http.Client{
			Transport: &http.Transport{
				DialContext: func(_ context.Context, network, addr string) (net.Conn, error) {
					return net.Dial(network, "example.com:443")
				},
				TLSClientConfig: &tls.Config{
					RootCAs:    certPool,
					ServerName: "example.com",
				},
			},
		}
	}

	if !testTrust(clientFactory, certPEM, certPriv) {
		log.Panic("failed to trust valid certificate")
	}

	if testTrust(clientFactory, wrongDomainPEM, wrongDomainPriv) {
		log.Panic("trusted invalid certificate")
	}

	if testTrust(clientFactory, wrongCAPEM, wrongCAPriv) {
		log.Panic("trusted invalid CA")
	}

	if testTrust(mitmClientFactory, certPEM, certPriv) {
		log.Panic("trusted system CA when it shouldn't")
	}

}
```

</details>

Reply:
Thanks for the example code. I've added a few test cases for the critical logic.